provider/source/src/units/trie.rs

```diff
@@ -0,0 +1,81 @@
+
+use std::collections::{BTreeMap, HashSet};
+
+use crate::cldr_serde;
+use crate::cldr_serde::units::info::ConvertUnit;
+use crate::SourceDataProvider;
+use icu::experimental::measure::provider::trie::{UnitsTrieV1, UnitsTrieV1Marker};
+use icu_provider::prelude::*;
+use zerotrie::ZeroTrieSimpleAscii;
+
+impl DataProvider<UnitsTrieV1Marker> for SourceDataProvider {
+    fn load(&self, _req: DataRequest) -> Result<DataResponse<UnitsTrieV1Marker>, DataError> {
+        self.check_req::<UnitsTrieV1Marker>(_req)?;
+
+        // Get all the constants in the form of a map from constant name to constant value as numerator and denominator.
+        let units_data: &cldr_serde::units::info::Resource = self
+            .cldr()?
+            .core()
+            .read_and_parse("supplemental/units.json")?;
+
+        // Get all the units and their conversion information.
+        let convert_units = &units_data.supplemental.convert_units.convert_units;
+
+        // Maps from unit id to its id which will be use to get the conversion information from the `UnitsInfoV1`.
+        let mut trie_map = BTreeMap::<Vec<u8>, usize>::new();
+
+        // Pre-process the conversion information to convert the factor and offset to scientific notation.
+        // This used to get the id for each unit which is used to get the conversion information from the `UnitsInfoV1`.
+        let mut convert_units_vec = Vec::<&ConvertUnit>::new();
+        for (unit_name, convert_unit) in convert_units {
+            let convert_unit_index = convert_units_vec.len();
+            convert_units_vec.push(convert_unit);
+            trie_map.insert(unit_name.as_bytes().to_vec(), convert_unit_index);
+        }
+
+        let units_conversion_trie = ZeroTrieSimpleAscii::try_from(&trie_map).map_err(|e| {
+            DataError::custom("Could not create ZeroTrie from units.json data")
+                .with_display_context(&e)
+        })?;
+
+        let result = UnitsTrieV1 {
+            trie: units_conversion_trie.convert_store(),
+        };
+
+        Ok(DataResponse {
+            metadata: Default::default(),
+            payload: DataPayload::from_owned(result),
+        })
+    }
+}
+
+impl crate::IterableDataProviderCached<UnitsTrieV1Marker> for SourceDataProvider {
+    fn iter_ids_cached(&self) -> Result<HashSet<DataIdentifierCow<'static>>, DataError> {
+        Ok(HashSet::from_iter([Default::default()]))
+    }
+}
+
+#[test]
+fn test_basic() {
+    use icu::locale::langid;
+    use icu_provider::prelude::*;
+
+    let provider = SourceDataProvider::new_testing();
+
+    let und: DataResponse<UnitsTrieV1Marker> = provider
+        .load(DataRequest {
+            id: DataIdentifierCow::from_locale(langid!("und").into()).as_borrowed(),
+            ..Default::default()
+        })
+        .unwrap();
+
+    let units_info = und.payload.get().to_owned();
+    let trie = &units_info.trie;
+
+    assert_eq!(trie.get("meter").unwrap(), 84);
```

Comment:
Testing that meter is a particular index is extremely brittle and will break at each CLDR update: [CLDR 45 diff](https://github.com/unicode-org/icu4x/pull/4782/files#diff-3db58552c2b814e04560154124f3cceb7d37d8bc200ea517a09082fb1f354e1d). Ideally the indices shouldn't even show in JSON data and be entirely an encoding detail.

Reply:
as you see in the TODO
```
    // TODO: add more tests that test the accessing of UnitsInfoV1Marker and get the correct info for each couple of units
```

I want in the next PR, to test that the id is pointing to the correct units info. After removing the duplication and use these data in the units converter. I will add these test cases.

Reply:
There is no guarantee to me as the reviewer that any TODO ever gets addressed. This is a blocking issue for me, because I will be the one who'll need to fix this during the next CLDR update if it's still there.

Either make this PR test that it's pointing to the correct units info, or remove this assertion.

Reply:
Okay, now I added a reasonable test cases that can proof why we need these ids.